File: UniBlock/src/main/scala/data/block/MerkleTree.scala
package it.unifi.nave.uniblock
package data.block

import data.event.Event

import it.unifi.nave.uniblock.helper.crypto.HashHelper

import scala.annotation.tailrec

object MerkleTree {

  def rootHash(events: List[Event]): String = getRootHash(events.map(_.hash))

  @tailrec
  private def getRootHash(hashes: List[String]): String = hashes match {
    case rootHash :: Nil => rootHash
    case _ => getRootHash(hashes.grouped(2).map(reduceHash).toList)
  }

  @tailrec
  private def reduceHash(hashes: List[String]): String = hashes match {
    case singleHash :: Nil => reduceHash(singleHash :: singleHash :: Nil)
    case firstHash :: secondHash => HashHelper.hash(Right(firstHash + secondHash))
  }

}

File: src/main/java/it/unifi/nave/uniblock/data/block/MerkleTree.java
package it.unifi.nave.uniblock.data.block;

import it.unifi.nave.uniblock.data.event.Event;
import it.unifi.nave.uniblock.helper.crypto.HashHelper;

import java.util.List;

public class MerkleTree {
    public static String rootHash(List<Event> events) {
        return getRootHash(events.stream().map(HashHelper::hash).toList());
    }

    private static String getRootHash(List<String> hashes) {
        if (hashes.size() == 1) {
            return hashes.get(0);
        } else {
            return null;
//            return getRootHash(Lists.partition(hashes, 2).stream().map(MerkleTree::reduceHash).toList());
        }
    }

    private static String reduceHash(List<String> hashes) {
        var first = hashes.get(0);
        var second = hashes.size() == 2 ? hashes.get(1) : first;
        return HashHelper.hash(first + second);
    }
}
